docs/fzsftp.man
.TH fzsftp 1 "January 2008" "" "FileZilla Manual"
.SH NAME
fzsftp \- SFTP connection handler of FileZilla
.SH DESCRIPTION
.B fzsftp
is part of FileZilla. It is used to handle SFTP (SSH File Transfer Protocol) connections. It usually gets called by FileZilla and is not intended to be used directly.
.P
.B fzsftp
is based on the psftp component of PuTTY.
.SH SUPPORT
Please visit http://filezilla\-project.org/ for further information. Report bugs only if you are using the latest version available from the FileZilla website.
.SH COPYRIGHT
Copyright (C) 2004\-2010  Tim Kosse
.br
Copyright (C) 1997\-2010  Simon Tatham and the PuTTY team
.P
FileZilla is distributed under the terms of the GNU General Public License version 2 or later.
.SH "SEE ALSO"
filezilla(1)


NEWS
This file only lists a short summary of the changes between FileZilla versions. For a full list of changes, please read the ChangeLog file.

Legend:
  ! Security related bugfix or otherwise critical bugfix 
  + New feature or important bugfix
  - Bugfixes

  Platform prefixes:
    MSW: Microsoft Windows
    *nix: Unix (e.g BSD) and Unix-like systems (e.g. Linux)
    OS X: Apple Mac OS X

3.3.2.1 (2010-03-21)

+ Remember location of update download directory
- Cancel file renaming on actions that change the file list, such as changing sort order or performing a refresh
- Cancel file renaming if starting a drag&drop operation
- Filter invalid characters if downloading through double-clicking files or if adding from the search dialog
- Fix site-specific bookmarks menu
- The menu did not reflect initial value of preserve timestamp option properly
- MSW: Keep FileZilla subdirectory in startmenu when updating, preserves custom order
- MSW: Add AppID support to installer

3.3.2 (2010-02-21)

+ *nix: Add support for forward and backward mouse buttons
- Speed limits can be re-enabled on SFTP connections
- Changed settings did not come into effect unless FileZilla was restarted
- Small performance improvements

3.3.2-rc2 (2010-02-14)

- Spelling fixes
- Rebuilt due to timer problems

3.3.2-rc1 (2010-02-13)

+ Speed limits can be quickly set using the menu and a new indicator in the status bar
+ Display number of bytes transferred and duration of transfer in log
+ MSW: React to back button of mouse in file lists
+ Updated blukis icon set
- Refactored large parts of the toolbar and menubar code
- A race condition in the socket code could lead to crashes
- Fix connecting to addresses longer than 15 characters if using SOCKS proxy
- Get add to queue in local directory tree context menu working again
- Do not open request dialogs if user is holding a mouse button
- Forward all Ctrl+Tab key events to tab bar
- Do not report SFTP exitcodes that do not originate from the server

3.3.1 (2010-01-03)

- Implement missing tab creation from context menu
- *nix: Reduce flicker on tab creation
- Set initial focus in chmod dialog to input box for numeric value
- Correct handling of unexpected closure of SFTP connection

3.3.1-rc1 (2009-12-27)

+ New tab context menu, can be used to close or refresh current tab or to close all but the current tab.
+ *nix, OS X: Display sort indicators in filelist header control
+ Remember size of dialog showing all files opened for editing
+ Add refresh item to filelist context menus
+ Scroll treeview horizontally if needed to keep item in view on changing directories
- Fix crash of fzsftp if server closes connection
- Successfully downloaded update file is no longer deleted if user cancels the finish page of the update check wizard
- Do not retry connecting if not trusting certificate on FTP over TLS connections
- Better reporting of handshake failures on FTP over TLS connections
- *nix: Check for xdg-open. This program from the xdg-utils is an indirect runtime dependency. It is used by wxWidgets to open URLs in the default browser
- Refactored most of the tabbing code
- Use proper address type on SOCKS proxies
- Add missing default button in bookmarks dialog

3.3.0.1 (2009-11-15)

+ Ctrl-Tab and Ctrl-Shift-Tab to cycle tabs now loop around if reaching the end of the tab list
- Fix crash on showing/hiding of directory trees after closing a tab
- Tabs no longer become unresponsive if not approving to close a busy tab
- Directory comparison did not work properly on all but the first tab
- Fix spurious error messages if connecting to a new site while synchronized browsing is enabled
- *nix: Work around a bug in KDE preventing restore from minimize to notification area
- Do not show update dialog if there is already some other dialog open
- Fix synchronized browsing option on site-specific bookmarks

3.3.0 (2009-11-09)

- Aborting a previous connection attempt by establishing a new connection in the same tab no longer renders the new connection unusable
- *nix: Quote filenames containing spaces if editing using default editor for text files
- *nix: Fix crash while emitting queue finished notifications

3.3.0-rc1 (2009-11-02)

+ *nix: Display desktop notification if all transfers have finished and FileZilla isn't focused
- Navigating controls inside tab using tabulator key now works on all tabs
- Fix splitter positions on horizontal interface layouts

3.3.0-beta2 (2009-10-29)

+ Give choice to abort old connection or to open a new tab when trying to connect while already connected
+ Double-click on empty space in tabbar opens new tab
+ Middle-click on tab closes it
+ Use splitter positions and filelist column setup of previous tab when opening new one
+ Make betas and release candidates check daily for new versions
- Fix crash when leaving settings dialog after tabs have been closed
- Fix crash on changing tabs if old tab has site-specific bookmarks but new one doesn't
- Layout of tabs no longer changes after leaving settings
- Fix issues with remembering last used server
- Don't use bold font for active tab
- Recursive operations now working in all tabs

3.3.0-beta1 (2009-10-27)

+ Tabbed interface to connect to multiple server at the same time
+ New welcome screen
+ Add Open to local directory tree context menu
- Fix mismatch of displayed and actually performed default file exists action
- MSW: Update mingwm10.dll to fix a crash at initialization
- MSW: FTP over TLS no longer crashes for users of VIA CPUs
- *nix: Fix crashes due to race condition in dbus code
- Fix SHA-512 calculating incorrect hashes on 64bit systems, used by SFTP and the update wizard

3.2.8.1 (2009-10-10)

- Fix crash if creating filters with attributes or permissions rule
- MSW: Fix column reordering and hiding in file lists
- MSW: Fix crash if trying to edit a file without any associated program nor default editor
- Use REST+STOR to resume upload instead of APPE if server advertises REST STREAM support


src/putty/fzsftp.c
#include "putty.h"
#include "misc.h"
#if defined(HAVE_GETTIMEOFDAY)
  #include <sys/time.h>
#elif defined(HAVE_FTIME)
  #include <sys/timeb.h>
#endif

int bytesAvailable[2] = { 0, 0 };

char* input_pushback = 0;

#ifndef _WINDOWS
#include <unistd.h>

char *input_buf = 0;
int input_buflen = 0, input_bufsize = 0;
#endif

static int ReadQuotas(int i)
{
#ifdef _WINDOWS
    HANDLE hin;
    DWORD savemode, newmode;

    hin = GetStdHandle(STD_INPUT_HANDLE);

    GetConsoleMode(hin, &savemode);
    newmode = savemode | ENABLE_PROCESSED_INPUT | ENABLE_LINE_INPUT;
    newmode &= ~ENABLE_ECHO_INPUT;
    SetConsoleMode(hin, newmode);

    while (bytesAvailable[i] == 0)
    {
	DWORD read;
	BOOL r;
	char buffer[21];

	r = ReadFile(hin, buffer, 20, &read, 0);
	if (!r || read == 0)
	    fatalbox("ReadFile failed in ReadQuotas");
	buffer[read] = 0;

	if (buffer[0] != '-')
	{
	    if (input_pushback != 0)
		fatalbox("input_pushback not null!");
	    else
		input_pushback = strdup(buffer);
	}
	else
	    ProcessQuotaCmd(buffer);
    }

    SetConsoleMode(hin, savemode);
#else
    char* line;
    while (bytesAvailable[i] == 0)
    {
	int error = 0;
	line = read_input_line(1, &error);
	if (line == NULL || error)
	{
	    fatalbox("read_input_line failed in ReadQuotas");
	    break;
	}

	if (line[0] != '-')
	{
	    if (input_pushback != 0)
		fatalbox("input_pushback not null!");
	    else
		input_pushback = strdup(line);
	}
	else
	    ProcessQuotaCmd(line);
	sfree(line);
    }
#endif //_WINDOWS
    return 1;
}

int RequestQuota(int i, int bytes)
{
    if (bytesAvailable[i] < -100)
	bytesAvailable[i] = 0;
    else if (bytesAvailable[i] < 0)
    {
	bytesAvailable[i]--;
	return bytes;
    }
    if (bytesAvailable[i] == 0)
    {
	bytesAvailable[i] = 0;
	fznotify(sftpUsedQuotaRecv + i);
    }
    if (bytesAvailable[i] == 0)
    {
	ReadQuotas(i);
    }

    if (bytesAvailable[i] < 0 || bytesAvailable[i] > bytes)
	return bytes;

    return bytesAvailable[i];
}

void UpdateQuota(int i, int bytes)
{
    if (bytesAvailable[i] < 0)
	return;

    if (bytesAvailable[i] > bytes)
	bytesAvailable[i] -= bytes;
    else
	bytesAvailable[i] = 0;
}

int ProcessQuotaCmd(const char* line)
{
    int direction = 0, number, pos;

    if (line[0] != '-')
	return 0;

    if (line[1] == '0')
	direction = 0;
    else if (line[1] == '1')
	direction = 1;
    else
	fatalbox("Invalid data received in ReadQuotas: Unknown direction");

    if (line[2] == '-')
    {
	bytesAvailable[direction] = -1;
	return 0;
    }

    number = 0;
    for (pos = 2;; pos++)
    {
	if (line[pos] == 0 || line[pos] == '\r' || line[pos] == '\n')
	    break;
	if (line[pos] < '0' || line[pos] > '9')
	    fatalbox("Invalid data received in ReadQuotas: Bytecount not a number");

	number *= 10;
	number += line[pos] - '0';
    }
    if (bytesAvailable[direction] == -1)
	bytesAvailable[direction] = number;
    else
	bytesAvailable[direction] += number;

    return 1;
}

char* get_input_pushback()
{
    char* pushback = input_pushback;
    input_pushback = 0;
    return pushback;
}

int has_input_pushback()
{
    if (input_pushback != 0)
	return 1;
    else
	return 0;
}

#ifndef _WINDOWS
static void clear_input_buffers(int free)
{
    if (free && input_buf != NULL)
	sfree(input_buf);
    input_buf = 0;
    input_bufsize = 0;
    input_buflen = 0;
}

char* read_input_line(int force, int* error)
{
    int ret;
    do {
	if (input_buflen >= input_bufsize) {
	    input_bufsize = input_buflen + 512;
	    input_buf = sresize(input_buf, input_bufsize, char);
	}
	ret = read(0, input_buf+input_buflen, 1);
	if (ret < 0) {
	    perror("read");
	    *error = 1;
	    clear_input_buffers(1);
	    return NULL;
	}
	if (ret == 0) {
	    /* eof on stdin; no error, but no answer either */
	    *error = 1;
	    clear_input_buffers(1);
	    return NULL;
	}

	if (input_buf[input_buflen++] == '\n') {
	    /* we have a full line */
	    char* buf = input_buf;
	    clear_input_buffers(0);
	    return buf;
	}
    } while(force);

    return NULL;
}
#endif

void fz_timer_init(_fztimer *timer)
{
#ifdef _WINDOWS
    timer->dwHighDateTime = timer->dwLowDateTime = 0;
#else
    timer->low = 0;
    timer->high = 0;
#endif
}

#ifndef _WINDOWS
// 1/10th of a second in microseconds
const static int notificationDelay = 1000000 / 10;
#else
// 1/10th of a second in 100 nanoseconds
const static unsigned int notificationDelay = 10000000 / 10;
#endif


int fz_timer_check(_fztimer *timer)
{
#ifdef _WINDOWS
    SYSTEMTIME sNow;
    FILETIME fNow;
    unsigned int diff;
    GetSystemTime(&sNow);
    SystemTimeToFileTime(&sNow, &fNow);

    diff = fNow.dwHighDateTime - timer->dwHighDateTime;
    if (!diff)
    {
	if ((fNow.dwLowDateTime - timer->dwLowDateTime) < notificationDelay)
	    return 0;
    }
    else if (diff == 1)
    {
	if (fNow.dwLowDateTime < timer->dwLowDateTime)
	{
	    if (((0xFFFFFFFFU - timer->dwLowDateTime) + fNow.dwLowDateTime) < notificationDelay)
		return 0;
	}
    }

    *timer = fNow;
#else
#if defined(HAVE_GETTIMEOFDAY)
    struct timeval tv;
    if (!gettimeofday(&tv, 0))
    {
	if (tv.tv_sec == timer->high)
	{
	    if ((tv.tv_usec - timer->low) < notificationDelay)
		return 0;
	}
	else if ((tv.tv_sec - timer->high) == 1)
	{
	    if (timer->low < tv.tv_usec)
	    {
		if (((1000000 - tv.tv_usec) + timer->low) < notificationDelay)
		    return 0;
	    }
	}

	timer->high = tv.tv_sec;
	timer->low = tv.tv_usec;
    }

#elif defined(HAVE_FTIME)
    struct timeb tp;

    ftime(&tp);
    if (tp.time == timer->high)
    {
	if ((tp.millitm - timer->low) < (notificationDelay / 1000))
	    return 0;
    }
    else if ((tp.time - timer->high) == 1)
    {
	if (timer->low < tp.millitm)
	{
	    if (((1000 - tp.millitm) + timer->low) < (notificationDelay / 1000))
		return 0;
	}
    }

    timer->high = tp.time;
    timer->low = tp.millitm;
#else
#error "Neither gettimeofday nor ftime available."
#endif
#endif
    return 1;
}
